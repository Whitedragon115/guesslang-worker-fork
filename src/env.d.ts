declare module "*.bin" {
  const src: ArrayBuffer;
  export default src;
}

declare module "__STATIC_CONTENT_MANIFEST" {
  const manifest: Record<string, unknown>;
  export default manifest;
}
