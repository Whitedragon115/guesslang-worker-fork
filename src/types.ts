import { ModelResult } from "@vscode/vscode-languagedetection";

export type DetectionOptions = {
  fineTune: boolean;
  verbose: boolean;
  expectedRelativeConfidence: number;
};

export type DetectionResult = {
  languageId: string;
  languageName: string;
  confidence: number;
  reliable: boolean;
  /**
   * Only available when `verbose` is `true`.
   */
  modelResults?: ModelResult[];
};
