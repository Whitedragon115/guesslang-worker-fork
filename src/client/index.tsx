import { render } from "hono/jsx/dom";
import { Home } from "./home";

const root = document.getElementById("root");
if (!root) {
  throw new Error("Root element not found");
}

render(<Home />, root);
