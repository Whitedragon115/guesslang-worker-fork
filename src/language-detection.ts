import NODE_WEIGHTS from "@vscode/vscode-languagedetection/model/group1-shard1of1.bin";
import NODE_MODEL_JSON from "@vscode/vscode-languagedetection/model/model.json";
import languagesMap from "./languages.json";
import { DetectionOptions, DetectionResult } from "./types";
import { ModelOperations, ModelResult } from "./vscode-languagedetection/index";

/*---------------------------------------------------------------------------------------------
 *  Ported from https://github.com/microsoft/vscode/blob/19ecb4b8337d0871f0a204853003a609d716b04e/src/vs/workbench/services/languageDetection/browser/languageDetectionSimpleWorker.ts
 *
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

const expectedRelativeConfidence = 0.2;
const positiveConfidenceCorrectionBucket1 = 0.05;
const positiveConfidenceCorrectionBucket2 = 0.025;
const negativeConfidenceCorrection = 0.5;

const adjustLanguageConfidence = (modelResult: ModelResult): ModelResult => {
  switch (modelResult.languageId) {
    // For the following languages, we increase the confidence because
    // these are commonly used languages in VS Code and supported
    // by the model.
    case "js":
    case "html":
    case "json":
    case "ts":
    case "css":
    case "py":
    case "xml":
    case "php":
      modelResult.confidence += positiveConfidenceCorrectionBucket1;
      break;
    // case 'yaml': // YAML has been know to cause incorrect language detection because the language is pretty simple. We don't want to increase the confidence for this.
    case "cpp":
    case "sh":
    case "java":
    case "cs":
    case "c":
      modelResult.confidence += positiveConfidenceCorrectionBucket2;
      break;

    // For the following languages, we need to be extra confident that the language is correct because
    // we've had issues like #131912 that caused incorrect guesses. To enforce this, we subtract the
    // negativeConfidenceCorrection from the confidence.

    // languages that are provided by default in VS Code
    case "bat":
    case "ini":
    case "makefile":
    case "sql":
    // languages that aren't provided by default in VS Code
    case "csv":
    case "toml":
      // Other considerations for negativeConfidenceCorrection that
      // aren't built in but suported by the model include:
      // * Assembly, TeX - These languages didn't have clear language modes in the community
      // * Markdown, Dockerfile - These languages are simple but they embed other languages
      modelResult.confidence -= negativeConfidenceCorrection;
      break;

    default:
      break;
  }
  return modelResult;
};

// See https://github.com/microsoft/vscode-languagedetection#advanced-options
const modelOperations = new ModelOperations({
  minContentSize: 0,
  modelJsonLoaderFunc: async () => {
    // return await import("@vscode/vscode-languagedetection/model/model.json");
    return NODE_MODEL_JSON;
  },
  weightsLoaderFunc: async () => {
    return NODE_WEIGHTS;
  },
});

const FALLBACK_LANGUAGE = {
  languageId: "txt",
  languageName: "Plain Text",
  confidence: 0,
  reliable: false,
} as const;

const DEFAULT_DETECTION_OPTIONS: DetectionOptions = {
  fineTune: true,
  verbose: false,
  expectedRelativeConfidence: expectedRelativeConfidence,
};

export const guessLang = async (text: string, fineTune: boolean) => {
  const modelResults = await modelOperations.runModel(text);
  const fineTunedModelResults = fineTune
    ? modelResults
        .map((modelResult) => adjustLanguageConfidence(modelResult))
        .sort((a, b) => b.confidence - a.confidence)
    : modelResults;
  return fineTunedModelResults;
};

export const detectLanguages = async (
  sampleContent: string,
  maybeOptions: Partial<DetectionOptions> = DEFAULT_DETECTION_OPTIONS,
): Promise<DetectionResult> => {
  const options = { ...DEFAULT_DETECTION_OPTIONS, ...maybeOptions };
  const modelResults = await guessLang(sampleContent, options.fineTune);
  const firstModelResult = modelResults.at(0);
  if (!firstModelResult) {
    return {
      ...FALLBACK_LANGUAGE,
      ...(options.verbose ? { modelResults: modelResults } : {}),
    };
  }

  return {
    languageId: firstModelResult.languageId,
    languageName:
      languagesMap[firstModelResult.languageId as keyof typeof languagesMap],
    confidence: firstModelResult.confidence,
    reliable: firstModelResult.confidence >= options.expectedRelativeConfidence,
    ...(options.verbose ? { modelResults: modelResults } : {}),
  };
};
