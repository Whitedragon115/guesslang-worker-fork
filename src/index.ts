import { Hono } from "hono";
import { serveStatic } from "hono/cloudflare-workers";
import { cors } from "hono/cors";
import { HTTPException } from "hono/http-exception";
import { Fragment } from "hono/jsx";
import { detectLanguages } from "./language-detection";
import { DetectionOptions } from "./types";
import manifestJSON from "__STATIC_CONTENT_MANIFEST";
import { Layout } from "./client/layout";

// See https://hono.dev/getting-started/cloudflare-workers
const app = new Hono();

app.use("/guess", cors());

app.get(
  "/static/*",
  serveStatic({
    root: "./",
    manifest: manifestJSON,
    rewriteRequestPath: (path) => path.replace(/^\/static/, ""),
  }),
);

app.get("/", (c) => {
  return c.html(Fragment({ children: Layout({}) }));
});

app.get("/guess", async (c) => {
  const query = c.req.query();
  const { text, ...options } = query;
  if (!text) {
    throw new HTTPException(400, {
      message: `Missing query parameter 'text'`,
    });
  }
  const guessResult = await detectLanguages(text, options);
  return c.json(guessResult);
});

app.post("/guess", async (c) => {
  const body = await c.req.json<
    Partial<DetectionOptions> & { text?: string }
  >();
  const { text, ...options } = body;
  if (!text) {
    throw new HTTPException(400, {
      message: `Missing query parameter 'text'`,
    });
  }
  const guessResult = await detectLanguages(text, options);
  return c.json(guessResult);
});

export default app;
