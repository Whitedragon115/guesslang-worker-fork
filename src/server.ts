import { serve } from '@hono/node-server'
import { Hono } from 'hono'
import { cors } from 'hono/cors'
import { HTTPException } from 'hono/http-exception'
import { Fragment } from 'hono/jsx'
import { serveStatic } from '@hono/node-server/serve-static'
import { Layout } from './client/layout.js'
import { detectLanguages } from './language-detection.js'
import { DetectionOptions } from './types.js'

const app = new Hono()

// Enable CORS for all routes
app.use('*', cors({
  origin: ['http://localhost:8787', 'http://127.0.0.1:8787'],
  allowMethods: ['GET', 'POST', 'OPTIONS'],
  allowHeaders: ['Content-Type', 'Authorization'],
}) as any)

// Serve static files
app.use('/static/*', serveStatic({
  root: './dist/static',
  rewriteRequestPath: (path) => path.replace(/^\/static/, ''),
}))

// Error handling middleware
app.onError((err, c) => {
  console.error('Server error:', err)
  
  if (err instanceof HTTPException) {
    return c.json({ error: err.message }, err.status)
  }
  
  return c.json({ error: 'Internal Server Error' }, 500)
})

app.get('/', (c) => {
  return c.html(Fragment({ children: Layout({}) }))
})

app.get('/guess', async (c) => {
  const query = c.req.query()
  const { text, ...options } = query

  if (!text) {
    throw new HTTPException(400, {
      message: 'Missing required parameter: text'
    })
  }

  try {
    const detectionOptions: Partial<DetectionOptions> = {
      fineTune: options.fineTune === 'true',
      verbose: options.verbose === 'true',
      expectedRelativeConfidence: options.expectedRelativeConfidence ? 
        parseFloat(options.expectedRelativeConfidence) : undefined
    }

    const result = await detectLanguages(text, detectionOptions)
    return c.json(result)
  } catch (error) {
    console.error('Language detection error:', error)
    throw new HTTPException(500, {
      message: 'Internal server error during language detection'
    })
  }
})

app.post('/guess', async (c) => {
  try {
    const body = await c.req.json()
    const { text, ...options } = body

    if (!text) {
      throw new HTTPException(400, {
        message: 'Missing required parameter: text'
      })
    }

    const result = await detectLanguages(text, options)
    return c.json(result)
  } catch (error) {
    if (error instanceof HTTPException) {
      throw error
    }
    
    console.error('Language detection error:', error)
    throw new HTTPException(500, {
      message: 'Internal server error during language detection'
    })
  }
})

// Health check endpoint
app.get('/health', (c) => {
  return c.json({ 
    status: 'healthy', 
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    memory: process.memoryUsage()
  })
})

const port = parseInt(process.env.PORT || '8787')

console.log(`Starting server on port ${port}`)

serve({
  fetch: app.fetch,
  port
})

console.log(`Server is running on http://localhost:${port}`)
