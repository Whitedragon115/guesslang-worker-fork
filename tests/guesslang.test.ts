import { detectLanguages, guessLang } from "../src/language-detection";
import { describe, expect, it } from "vitest";

describe("guessLang", () => {
  it("should match snapshot for the given code snippet", async () => {
    // Example from https://github.com/microsoft/vscode-languagedetection
    const result = await guessLang(
      `
  function makeThing(): Thing {
      let size = 0;
      return {
          get size(): number {
          return size;
          },
          set size(value: string | number | boolean) {
          let num = Number(value);
          // Don't allow NaN and stuff.
          if (!Number.isFinite(num)) {
              size = 0;
              return;
          }
          size = num;
          },
      };
  }
`,
      false,
    );
    expect(result).toMatchSnapshot();
  });

  it("should match snapshot for the given code snippet", async () => {
    // Example from https://github.com/microsoft/vscode-languagedetection
    const result = await guessLang(
      `
  function makeThing(): Thing {
      let size = 0;
      return {
          get size(): number {
          return size;
          },
          set size(value: string | number | boolean) {
          let num = Number(value);
          // Don't allow NaN and stuff.
          if (!Number.isFinite(num)) {
              size = 0;
              return;
          }
          size = num;
          },
      };
  }
`,
      true,
    );
    expect(result).toMatchSnapshot();
  });
});

describe("detectLanguages", () => {
  it("should return the correct language for the given code snippet", async () => {
    // Example from https://github.com/microsoft/vscode-languagedetection
    const result = await detectLanguages(`
  function makeThing(): Thing {
      let size = 0;
      return {
          get size(): number {
          return size;
          },
          set size(value: string | number | boolean) {
          let num = Number(value);
          // Don't allow NaN and stuff.
          if (!Number.isFinite(num)) {
              size = 0;
              return;
          }
          size = num;
          },
      };
  }
`);
    expect(result).toMatchObject({
      confidence: 0.41006417870521544,
      isCreditable: true,
      languageId: "ts",
      languageName: "TypeScript",
    });
  });

  it("should return fallback txt for empty input", async () => {
    // Example from https://github.com/microsoft/vscode-languagedetection
    const result = await detectLanguages(``);
    expect(result).toMatchObject({
      confidence: 0,
      isCreditable: false,
      languageId: "txt",
      languageName: "Plain Text",
    });
  });
});
