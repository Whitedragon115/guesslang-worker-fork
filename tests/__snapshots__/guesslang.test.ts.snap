// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`guessLang > should match snapshot for the given code snippet 1`] = `
[
  {
    "confidence": 0.36006417870521545,
    "languageId": "ts",
  },
  {
    "confidence": 0.12616758048534393,
    "languageId": "js",
  },
  {
    "confidence": 0.11067066341638565,
    "languageId": "rs",
  },
  {
    "confidence": 0.049941033124923706,
    "languageId": "cs",
  },
  {
    "confidence": 0.04073625057935715,
    "languageId": "html",
  },
  {
    "confidence": 0.034685082733631134,
    "languageId": "dart",
  },
  {
    "confidence": 0.028705749660730362,
    "languageId": "c",
  },
  {
    "confidence": 0.02582576498389244,
    "languageId": "lua",
  },
  {
    "confidence": 0.019651010632514954,
    "languageId": "cpp",
  },
  {
    "confidence": 0.019124053418636322,
    "languageId": "mm",
  },
  {
    "confidence": 0.01554538682103157,
    "languageId": "pm",
  },
  {
    "confidence": 0.015271097421646118,
    "languageId": "ps1",
  },
  {
    "confidence": 0.012661250308156013,
    "languageId": "java",
  },
  {
    "confidence": 0.012522794306278229,
    "languageId": "groovy",
  },
  {
    "confidence": 0.012012877501547337,
    "languageId": "swift",
  },
  {
    "confidence": 0.010405079461634159,
    "languageId": "pas",
  },
  {
    "confidence": 0.009881828911602497,
    "languageId": "md",
  },
  {
    "confidence": 0.008674976415932178,
    "languageId": "prolog",
  },
  {
    "confidence": 0.00856003351509571,
    "languageId": "php",
  },
  {
    "confidence": 0.008517211303114891,
    "languageId": "sql",
  },
  {
    "confidence": 0.008150268346071243,
    "languageId": "ml",
  },
  {
    "confidence": 0.006514098960906267,
    "languageId": "r",
  },
  {
    "confidence": 0.005204849410802126,
    "languageId": "go",
  },
  {
    "confidence": 0.004512660205364227,
    "languageId": "matlab",
  },
  {
    "confidence": 0.004238471854478121,
    "languageId": "jl",
  },
  {
    "confidence": 0.003899412229657173,
    "languageId": "erl",
  },
  {
    "confidence": 0.003883998841047287,
    "languageId": "tex",
  },
  {
    "confidence": 0.0036005694419145584,
    "languageId": "css",
  },
  {
    "confidence": 0.0031094681471586227,
    "languageId": "asm",
  },
  {
    "confidence": 0.0027291469741612673,
    "languageId": "kt",
  },
  {
    "confidence": 0.0026916172355413437,
    "languageId": "py",
  },
  {
    "confidence": 0.0023645830806344748,
    "languageId": "xml",
  },
  {
    "confidence": 0.00209893099963665,
    "languageId": "v",
  },
  {
    "confidence": 0.002020909683778882,
    "languageId": "hs",
  },
  {
    "confidence": 0.001936757704243064,
    "languageId": "f90",
  },
  {
    "confidence": 0.0015712525928393006,
    "languageId": "scala",
  },
  {
    "confidence": 0.0014332145219668746,
    "languageId": "dm",
  },
  {
    "confidence": 0.001254789880476892,
    "languageId": "rb",
  },
  {
    "confidence": 0.0012241607764735818,
    "languageId": "bat",
  },
  {
    "confidence": 0.0011867597931995988,
    "languageId": "coffee",
  },
  {
    "confidence": 0.0011775789316743612,
    "languageId": "sh",
  },
  {
    "confidence": 0.001136972103267908,
    "languageId": "vba",
  },
  {
    "confidence": 0.000994101632386446,
    "languageId": "ex",
  },
  {
    "confidence": 0.0009934992995113134,
    "languageId": "cbl",
  },
  {
    "confidence": 0.0006263475515879691,
    "languageId": "csv",
  },
  {
    "confidence": 0.00045713630970567465,
    "languageId": "yaml",
  },
  {
    "confidence": 0.000399948243284598,
    "languageId": "lisp",
  },
  {
    "confidence": 0.0003188799018971622,
    "languageId": "clj",
  },
  {
    "confidence": 0.00022272310161497444,
    "languageId": "ini",
  },
  {
    "confidence": 0.00021139078307896852,
    "languageId": "cmake",
  },
  {
    "confidence": 0.00016815742128528655,
    "languageId": "json",
  },
  {
    "confidence": 0.000023193719243863598,
    "languageId": "toml",
  },
  {
    "confidence": 0.000014289931641542353,
    "languageId": "dockerfile",
  },
  {
    "confidence": 0.000005981031335977605,
    "languageId": "makefile",
  },
]
`;

exports[`guessLang > should match snapshot for the given code snippet 2`] = `
[
  {
    "confidence": 0.41006417870521544,
    "languageId": "ts",
  },
  {
    "confidence": 0.17616758048534392,
    "languageId": "js",
  },
  {
    "confidence": 0.11067066341638565,
    "languageId": "rs",
  },
  {
    "confidence": 0.09073625057935715,
    "languageId": "html",
  },
  {
    "confidence": 0.0749410331249237,
    "languageId": "cs",
  },
  {
    "confidence": 0.058560033515095714,
    "languageId": "php",
  },
  {
    "confidence": 0.05370574966073036,
    "languageId": "c",
  },
  {
    "confidence": 0.05360056944191456,
    "languageId": "css",
  },
  {
    "confidence": 0.052691617235541346,
    "languageId": "py",
  },
  {
    "confidence": 0.05236458308063448,
    "languageId": "xml",
  },
  {
    "confidence": 0.05016815742128529,
    "languageId": "json",
  },
  {
    "confidence": 0.044651010632514955,
    "languageId": "cpp",
  },
  {
    "confidence": 0.037661250308156015,
    "languageId": "java",
  },
  {
    "confidence": 0.034685082733631134,
    "languageId": "dart",
  },
  {
    "confidence": 0.026177578931674363,
    "languageId": "sh",
  },
  {
    "confidence": 0.02582576498389244,
    "languageId": "lua",
  },
  {
    "confidence": 0.019124053418636322,
    "languageId": "mm",
  },
  {
    "confidence": 0.01554538682103157,
    "languageId": "pm",
  },
  {
    "confidence": 0.015271097421646118,
    "languageId": "ps1",
  },
  {
    "confidence": 0.012522794306278229,
    "languageId": "groovy",
  },
  {
    "confidence": 0.012012877501547337,
    "languageId": "swift",
  },
  {
    "confidence": 0.010405079461634159,
    "languageId": "pas",
  },
  {
    "confidence": 0.009881828911602497,
    "languageId": "md",
  },
  {
    "confidence": 0.008674976415932178,
    "languageId": "prolog",
  },
  {
    "confidence": 0.008150268346071243,
    "languageId": "ml",
  },
  {
    "confidence": 0.006514098960906267,
    "languageId": "r",
  },
  {
    "confidence": 0.005204849410802126,
    "languageId": "go",
  },
  {
    "confidence": 0.004512660205364227,
    "languageId": "matlab",
  },
  {
    "confidence": 0.004238471854478121,
    "languageId": "jl",
  },
  {
    "confidence": 0.003899412229657173,
    "languageId": "erl",
  },
  {
    "confidence": 0.003883998841047287,
    "languageId": "tex",
  },
  {
    "confidence": 0.0031094681471586227,
    "languageId": "asm",
  },
  {
    "confidence": 0.0027291469741612673,
    "languageId": "kt",
  },
  {
    "confidence": 0.00209893099963665,
    "languageId": "v",
  },
  {
    "confidence": 0.002020909683778882,
    "languageId": "hs",
  },
  {
    "confidence": 0.001936757704243064,
    "languageId": "f90",
  },
  {
    "confidence": 0.0015712525928393006,
    "languageId": "scala",
  },
  {
    "confidence": 0.0014332145219668746,
    "languageId": "dm",
  },
  {
    "confidence": 0.001254789880476892,
    "languageId": "rb",
  },
  {
    "confidence": 0.0011867597931995988,
    "languageId": "coffee",
  },
  {
    "confidence": 0.001136972103267908,
    "languageId": "vba",
  },
  {
    "confidence": 0.000994101632386446,
    "languageId": "ex",
  },
  {
    "confidence": 0.0009934992995113134,
    "languageId": "cbl",
  },
  {
    "confidence": 0.00045713630970567465,
    "languageId": "yaml",
  },
  {
    "confidence": 0.000399948243284598,
    "languageId": "lisp",
  },
  {
    "confidence": 0.0003188799018971622,
    "languageId": "clj",
  },
  {
    "confidence": 0.00021139078307896852,
    "languageId": "cmake",
  },
  {
    "confidence": 0.000014289931641542353,
    "languageId": "dockerfile",
  },
  {
    "confidence": -0.4914827886968851,
    "languageId": "sql",
  },
  {
    "confidence": -0.4987758392235264,
    "languageId": "bat",
  },
  {
    "confidence": -0.49937365244841203,
    "languageId": "csv",
  },
  {
    "confidence": -0.499777276898385,
    "languageId": "ini",
  },
  {
    "confidence": -0.49997680628075614,
    "languageId": "toml",
  },
  {
    "confidence": -0.499994018968664,
    "languageId": "makefile",
  },
]
`;
